src/main/java/hudson/tasks/junit/pipeline/JUnitResultsStep.java

```diff
@@ -0,0 +1,160 @@
+
+    /**
+     * {@link FileSet} "includes" string, like "foo/bar/*.xml"
+     */
+    private final String testResults;
+
+    /**
+     * If true, retain a suite's complete stdout/stderr even if this is huge and the suite passed.
+     * @since 1.358
+     */
+    private boolean keepLongStdio;
+
+    /**
+     * {@link TestDataPublisher}s configured for this archiver, to process the recorded data.
+     * For compatibility reasons, can be null.
+     * @since 1.320
+     */
+    private DescribableList<TestDataPublisher, Descriptor<TestDataPublisher>> testDataPublishers;
+
+    private Double healthScaleFactor;
+
+    /**
+     * If true, don't throw exception on missing test results or no files found.
+     */
+    private boolean allowEmptyResults;
+
+    @DataBoundConstructor
+    public JUnitResultsStep(String testResults) {
+        this.testResults = testResults;
+    }
+
+    public String getTestResults() {
+        return testResults;
+    }
+
+    public double getHealthScaleFactor() {
+        return healthScaleFactor == null ? 1.0 : healthScaleFactor;
+    }
+
+    /**
+     * @param healthScaleFactor Health scale factor.
+     *
+     * @since 1.2-beta-1
+     */
+    @DataBoundSetter
+    public final void setHealthScaleFactor(double healthScaleFactor) {
+        this.healthScaleFactor = Math.max(0.0, healthScaleFactor);
+    }
+
+    public @Nonnull
+    List<TestDataPublisher> getTestDataPublishers() {
+        return testDataPublishers == null ? Collections.<TestDataPublisher>emptyList() : testDataPublishers;
+    }
+
+    /**
+     * @param testDataPublishers Test data publishers.
+     *
+     * @since 1.2
+     */
+    @DataBoundSetter public final void setTestDataPublishers(@Nonnull List<TestDataPublisher> testDataPublishers) {
+        this.testDataPublishers = new DescribableList<TestDataPublisher,Descriptor<TestDataPublisher>>(Saveable.NOOP);
+        this.testDataPublishers.addAll(testDataPublishers);
+    }
+
+    /**
+     * @return the keepLongStdio.
+     */
+    public boolean isKeepLongStdio() {
+        return keepLongStdio;
+    }
+
+    /**
+     * @param keepLongStdio Whether to keep long stdio.
+     *
+     * @since 1.2-beta-1
+     */
+    @DataBoundSetter public final void setKeepLongStdio(boolean keepLongStdio) {
+        this.keepLongStdio = keepLongStdio;
+    }
+
+    /**
+     *
+     * @return the allowEmptyResults
+     */
+    public boolean isAllowEmptyResults() {
+        return allowEmptyResults;
+    }
+
+    @DataBoundSetter public final void setAllowEmptyResults(boolean allowEmptyResults) {
+        this.allowEmptyResults = allowEmptyResults;
+    }
+
+    @Override
+    public StepExecution start(StepContext context) throws Exception {
+        return new JUnitResultsStepExecution(this, context);
+    }
+
+    @Extension
+    public static class DescriptorImpl extends StepDescriptor {
+        @Override
+        public String getFunctionName() {
+            return "junitResults";
```

Comment:
*nod* I'm not 100% sure how we can test it such that `junit` means different things in a single run (albeit with a restart) but if we can find a way to do that, I'm all in favor.